File: src/main/java/mvp/UserContract.java
package mvp;

import java.util.List;

import chapter2OOP.User;

public interface UserContract {

    interface UserView extends BaseContract.IView {

        void displayUser(List<User> users);

        void displayError(String message);

    }

    interface UserPresenter extends BaseContract.IPresenter<UserView> {

        void fetchUsers();


    }


    interface UserInteractor extends BaseContract.IInteractor<UserOutput> {
        void fetchUsersFromNetwork();
    }

    interface UserOutput extends BaseContract.IOutput {

        void onUsersFetched(List<User> users);

        void onErrorOccurred(String message);
    }

}

File: src/main/kotlin/chapter2OOP/mvp/UserContract.kt
package chapter2OOP.mvp

import chapter2OOP.User
import chapter2OOP.mvp.BaseContract.*

interface UserContract {

    interface UserView : IView {
        fun displayUser(users: List<User>)
        fun displayError(message: String?)
    }

    interface UserPresenter : IPresenter<UserView> {
        fun fetchUsers()
    }

    interface UserInteractor : IInteractor {
        fun fetchUsersFromNetwork()
    }

    interface UserOutput : IOutput {
        fun onUsersFetched(users: List<User>)
        fun onErrorOccurred(message: String)
    }
}